Adapt for Lagom 1.5.0 deployment

renatocaval opened the pull request:
This is the PR to adapt the guide for Lagom 1.5.0. I few things to notice:

* I explicitly left Lagom 1.4.0 out although we said that it could be a good idea to keep it (more on this further down)
* we don't have it the guide for akka, but I understand that some work was already done to reuse some content. I tried to keep the same approach and only copy-n-paste text when it was really necessary to add some Lagom specifics (or remove things)
* I remove some text that just because I found it hard to reconcile with some new sentences. For instance, the Lagom part won't have the sentence: "For the most part, we'll rely on the defaults, for example, the default port that Akka remoting binds to is 2552. ". This is because, I found it more useful to tell the users that most of the configuration is already handled by Lagom and that they should not worry about it.

### About Lagom 1.4.0

I explicitly left 1.4.0 out because I have the impression that it will only add confusion to the guide. It's better to have a straight guide dedicated to 1.5.0 instead of here and there a note like, "but if you are using 1.4.0, you need that extra bit". 

On top of that, I also don't think that we want to maintain two other branches of the samples "scala/1.4.x" and "java/1.4.x". That's already too much for us.

Instead, I would like to suggest the following:

* let's keep 1.4.0 out for the moment
* if we decide to re-add it, let's add an "Appendix C: Deploying Lagom 1.4.x" that will only explain the necessary bits to enable Akka Management and Cluster Bootstrap in Lagom 1.4.x. Once that in place, the rest of the guide should be valid for both versions.
* let's not maintain samples for 1.4.x. If someone needs to do the same in Lagom 1.4.x, they can get inspiration from the 1.5.x branch and the above mention appendix.

jroper commented on src/main/paradox/includes/forming-a-cluster.md:
```diff
@@ -65,42 +67,65 @@ akka {
+        # the actor system will be terminated
         shutdown-after-unsuccessful-join-seed-nodes = 60s
+        # exit jvm on actor system termination
+        akka.coordinated-shutdown.exit-jvm = on
```
This is in the wrong place I think, it should be at the top level, not under cluster.

renatocaval replied:
yes, of course

jroper commented on src/main/paradox/includes/forming-a-cluster.md:
````diff
@@ -65,42 +67,65 @@ akka {
-```
-
-This will expose liveness and readiness health checks on `/alive` and `/ready` respectively. 
+Akka management HTTP includes [health check routes](https://developer.lightbend.com/docs/akka-management/current/healthchecks.html) that will expose liveness and readiness health checks on `/alive` and `/ready` respectively. 
````
Since there's now no longer anything to configure here, should we move this discussion down to the section on configuring the health check probes in the k8s spec? It's just a bit odd because we introduce this subject, then go on to talk about other things, then come back to it. I think it would flow better if we only talked about health checks in one place in this page.

jroper commented:
The CI failure here appears unrelated, the build got through deploying and testing Scala, Java with sbt, and almost all the way through Java with Maven, but then got this error when creating the inventory application secret:

```
Error from server (Forbidden): secrets "inventory-application-secret" is forbidden: caches not synchronized
```

We can probably put it down to a quirk of running Minishift in a resource constrained environment.

renatocaval commented:
@jroper, I implemented the requested changes and added a few more improvements like links to the health-check section. 

Is ready for a second review round. Thanks!

TimMoore commented on src/main/paradox/includes/forming-a-cluster.md:
```diff
@@ -53,54 +53,70 @@ Maven
-The first thing that's needed is a general Akka cluster configuration. For the most part, we'll rely on the defaults, for example, the default port that Akka remoting binds to is 2552. But there are a few things we need to tweak. We need to first enable Akka cluster by making it the Actor provider. We also want to tell Akka to shut itself down if it's unable to join the cluster after a given time ...
-
+The first thing that's needed is a general Akka cluster configuration. For the most part, we'll rely on the defaults, for example, the default port that Akka remoting binds to is 2552. But there are a few things we need to tweak. We need to first enable Akka cluster by making it the Actor provider. We also want to tell Akka to shut itself down if it's unable to join the cluster after a given time ...
+Setup](setup/index.md)
```
Oh hang on... clicked too soon. Looks like a typo here.

renatocaval replied:
Oeps!

ignasi35 commented on src/main/paradox/lagom/index.md:
```diff
@@ -1,8 +1,8 @@
 # Deploying a Lagom application to OpenShift
 
-For this guide we will be using the Lagom shopping cart sample application. It is available for both Java and Scala, and uses a PostgreSQL database.
+For this guide we will be using the Lagom shopping cart sample application. It is available for both Java and Scala, and uses a PostgreSQL database. It covers how to deploy a Lagom 1.5.0 application.
```
```suggestion
For this guide we will be using the Lagom shopping cart sample application. It is available for both [Java](https://github.com/lagom/shopping-cart-java/tree/1.5.x) and [Scala](https://github.com/lagom/shopping-cart-scala/tree/1.5.x), and uses a PostgreSQL database. This guide covers how to deploy a Lagom 1.5.0 application.
```

ignasi35 replied:
Added links to the sample apps and removed a pronoun.

ignasi35 commented on src/main/paradox/includes/forming-a-cluster.md:
```diff
@@ -53,54 +53,70 @@ Maven
 
     cluster {
+        # after 60s of unsuccessul attempts to form a cluster, 
+        # the actor system will be terminated
```
```suggestion
        # the actor system will shut down
```

ignasi35 commented on src/main/paradox/lagom/forming-a-cluster.md:
```diff
@@ -6,29 +6,34 @@ If you're using any of the Akka cluster based features of Lagom, such as Lagom P
-## Starting
+These components are automatically added by Lagom whenever Lagom Persistence or Lagom Pub Sub is in use, or when you explicitly enable clustering in Lagom. The following sections explain their roles, how they fit together and which extra configurations are needed.
+
+### Akka Cluster
```
Mental note: maintaining this section (`Akka Clsuter`) which is duplicate and customized for vanilla Akka in `includes` or particularly for Lagom in this `lagom/forming-a-cluster-md` page will be error prone. Meh.

renatocaval replied:
Maybe, but I this was already designed to be re-usable. I didn't want to change that now. When we add an Akka guide we can revisit that approach.

ignasi35 commented on src/main/paradox/lagom/forming-a-cluster.md:
````diff
@@ -6,29 +6,34 @@ If you're using any of the Akka cluster based features of Lagom, such as Lagom P
-### Scala
+```HOCON
+# after 60s of unsuccessul attempts to form a cluster, 
+# the actor system will be terminated
````
```suggestion
# the actor system will shut down
```

ignasi35 commented on src/main/paradox/includes/forming-a-cluster.md:
```diff
@@ -53,54 +53,70 @@ Maven
 Hence, we need to use a different discovery method for cluster bootstrap, and for Kubernetes, the simplest method is to use the Kubernetes API, which will return all nodes regardless of their readiness state. 
 
-First, you need to depend on `akka-discovery-kubernetes-api`  (see the list of dependencies listed at the beginning of this page for the exact syntax) and then you must configure that discovery mechanism like so:
+First, you need to depend on `akka-discovery-kubernetes-api`: 
```
👍🏼 This is an OpenShift-specific guide. No need to mention other methods.

ignasi35 commented on src/main/paradox/includes/forming-a-cluster.md:
```diff
@@ -219,9 +235,19 @@ ports:
 
+Akka management HTTP includes [health check routes](https://developer.lightbend.com/docs/akka-management/current/healthchecks.html) that will expose liveness and readiness health checks on `/alive` and `/ready` respectively. 
+
+In Kubernetes, if an application is live, it means its running - it hasn't crashed. But it may not necessarily be ready to serve requests, for example, it might not yet have managed to connect to a database, or in our case, it may not have formed a cluster yet. By separating liveness and readiness, Kubernetes can distinguish between fatal errors, like crashing, and transient errors, like not bein ...
```
```suggestion
In Kubernetes, if an application is live, it means it is running - it hasn't crashed. But it may not necessarily be ready to serve requests, for example, it might not yet have managed to connect to a database, or, in our case, it may not have formed a cluster yet. By separating liveness and readiness, Kubernetes can distinguish between fatal errors, like crashing, and transient errors, like not being able to contact other resources that the application depends on, allowing Kubernetes to make more intelligent decisions about whether an application needs to be restarted, or if it just needs to be given time to sort itself out.
```

ignasi35 replied:
(the above suggestion is a comma and a grammatical `its`->`it is`)

Bonus: I found this paragraph too hard to read: 2 very long sentences.

ignasi35 replied:
not sure where in the paragraph but we could also link to https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes

renatocaval replied:
I just added a break line and links, let me know if that does the trick.

ignasi35 commented:
IMHO, only the conflict is blocking the merge.

Nice changes! thanks @renatocaval
